asf_heat_pump_suitability/analysis/flats_on_fossils/analysis_fill_building_rise_methods.py

```diff
@@ -606,11 +567,287 @@ def extend_df_building_rise_from_thresholds(
+        )
+        X_test = model_df.filter(pl.col("building_id").is_in(test_groups)).select(
+            ["height", "property_per_m2"]
+        )
+
+        y_train = model_df.filter(pl.col("building_id").is_in(train_groups))[
+            "FLAT_STOREY_COUNT"
+        ]
+        y_test = model_df.filter(pl.col("building_id").is_in(test_groups))[
+            "FLAT_STOREY_COUNT"
+        ]
+
+        # Train model
+        reg = linear_model.LinearRegression().fit(X_train, y_train)
+
+        # R2 on test data
+        test_r2.append(reg.score(X_test, y_test))
+
+    avg_test_r2.append(np.mean(test_r2))
+
+print(avg_test_r2)
+print(np.mean(avg_test_r2))
+print(np.median(avg_test_r2))
+
+# %% [markdown]
+# ## Train and evaluate final selected model
+#
+# We train the final model on the full dataset. We're not too worried about overfitting because it's a linear model with 2 features.
+#
+# In the evaluation below we classify our actual and predicted storey counts into low-, medium-, and high-rise buildings and compare them. We can see that we pretty accurately predict the number of high-rises, we under-predict the number of low-rises, and over-predict the number of high-rises. We can see that our model accuracy is around 80%.
+#
+# Comparatively, the polynomial regression model has a much more accurate prediction of the number of medium-, and low-rises. However, the model accuracy is still around 80% so for our use case where we care about the accuracy of individual predictions, it's not necessarily a more useful model. It's also much less intuitive and explainable, especially when viewing the coefficients.
+#
+# When viewing our evaluation results, we should be cognizant of the fact that our ground truth y values have known errors. In the random sample of 15 flats we took, we found 3/15 had erroneous flat storey count data when compared to google maps, equivalent to 80%. So although our model shows inaccurate predictions for 20% of buildings in our labelled dataset, it could be the case that some of these labels are incorrect.
+
+# %%
+# Prepare model data subset
+model_df = (
+    flats_epc_df.filter(
+        pl.col("FLAT_STOREY_COUNT").is_not_null(),
+        pl.col("height").is_not_null(),
+        pl.col("property_per_m2").is_not_null(),
+    )
+    .with_columns(
+        (pl.col("height") / pl.col("FLAT_STOREY_COUNT")).alias("meters_per_storey")
+    )
+    .filter((pl.col("COUNTRY") != "Scotland"), (pl.col("height") >= 2))
+)
+
+uprn_sample = model_df["UPRN"].to_list()
+X = model_df.select(["height", "property_per_m2"])
+y = model_df["FLAT_STOREY_COUNT"]
+
+X_train, X_test, y_train, y_test = train_test_split(
+    X, y, test_size=0.25, random_state=1
+)
+
+# Train final candidate model
+reg = linear_model.LinearRegression().fit(X, y)
+
+# %%
+reg.score(X, y)
+
+# %%
+# Get full data predictions of final candidate model
+candidate_preds = reg.predict(X)
+residuals = y - candidate_preds
+
+plt.hist(residuals, bins=100)
```

Comment:
Add range to make plot clearer
```suggestion
plt.hist(residuals, bins=100)
plt.xlim(-15, 15)
```